dialogue-core/src/main/java/com/palantir/dialogue/core/DeprecationWarningChannel.java

```diff
@@ -37,42 +38,47 @@
  * be used to understand more granular rates of deprecated calls against a particular service using the
  * {@code service-name} tag.
  */
-final class DeprecationWarningChannel implements Channel {
+final class DeprecationWarningChannel implements ChannelEndpointStage {
     private static final Logger log = LoggerFactory.getLogger(DeprecationWarningChannel.class);
     private static final Object SENTINEL = new Object();
 
-    private final Channel delegate;
+    private final ChannelEndpointStage delegate;
     private final ClientMetrics metrics;
     private final Cache<String, Object> loggingRateLimiter =
             Caffeine.newBuilder().expireAfterWrite(Duration.ofMinutes(1)).build();
 
-    DeprecationWarningChannel(Channel delegate, TaggedMetricRegistry metrics) {
+    DeprecationWarningChannel(ChannelEndpointStage delegate, TaggedMetricRegistry metrics) {
         this.delegate = delegate;
         this.metrics = ClientMetrics.of(metrics);
     }
 
     @Override
-    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
-        return DialogueFutures.addDirectCallback(
-                delegate.execute(endpoint, request), DialogueFutures.onSuccess(response -> {
-                    if (response != null) {
-                        response.getFirstHeader("deprecation").ifPresent(deprecated -> {
-                            metrics.deprecations(endpoint.serviceName()).mark();
-                            if (tryAcquire(endpoint.serviceName())) {
-                                log.warn(
-                                        "Using a deprecated endpoint when connecting to service",
-                                        SafeArg.of("serviceName", endpoint.serviceName()),
-                                        SafeArg.of("endpointHttpMethod", endpoint.httpMethod()),
-                                        SafeArg.of("endpointName", endpoint.endpointName()),
-                                        SafeArg.of("endpointClientVersion", endpoint.version()),
-                                        SafeArg.of(
-                                                "service",
-                                                response.getFirstHeader("server")
-                                                        .orElse("no server header provided")));
-                            }
-                        });
+    public EndpointChannel endpoint(Endpoint endpoint) {
+        EndpointChannel proceed = delegate.endpoint(endpoint);
+        FutureCallback<Response> callback = createCallback(endpoint);
+        return req -> DialogueFutures.addDirectCallback(proceed.execute(req), callback);
```

Comment:
super minor nit: may want to use something concrete with a toString for debugging.

Reply:
Yep I'm very much aware that I added all these nice toStrings and now lambdas & anonymous classes will have ruined them all.